doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+
+### Use as a part of 2FA (OTP)
+
+When 2FA will be enabled, it will work (and be required) only for the user's
+primary password, not for any of the app passwords, as that would cause too much
+trouble.
+
+### Use as a fallback mechanism when 2FA is activated, but the client app does not support it
+
+As 2FA will not be applied to app passwords, the user can use an app password to
+log into a service that does not support 2FA.
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+App passwords will be stored in a new LDAP subtree in the form
+```
+dn: uid=user-1,cn=myapp,cn=apps,cn=accounts,$SUFFIX
+uid: user-1
+displayName: My eMail app password for tablet
+appname: myapp
+entryUUID: XXXXXXXXX-XXXXXXXXXX-XXXXXXXXXX
+userPassword: *********************
+managedby: uid=user,cn=users,cn=accounts,$SUFFIX
+
+dn: uid=user-2,cn=myapp,cn=apps,cn=accounts,$SUFFIX
+uid: user-2
+displayName: My eMail app password for smartphone
+appname: myapp
+entryUUID: XXXXXXXXX-XXXXXXXXXX-XXXXXXXXXX
+userPassword: *********************
+managedby: uid=user,cn=users,cn=accounts,$SUFFIX
+```
+Every app will have its own subtree in which users will be able to create their
+objects if they will want to use that app - this means objects containing app
+passwords will not be grouped by users, but by apps. The object containing an
+app password will have the following attributes:
+* `uid` - will identify the object, and will also contain a number - this allows
+  for one user having multiple passwords for the same app, its value will be
+  enforced by LDAP ACIs
+* `displayname` - the descriptive label of the password, search display
+  attribute
+* `appname` - the name of the app the password is used for, it is the same as
+  the CN of the parent entry of the object
+* `entryUUID` - a UUID, which will be useful to uniquely identify a password
+  even in extreme cases when `uid`, `displayname` and `appname` will be the same
+  for some password; `entryUUID` will be also used to identify the app password
+  to delete
+* `userpassword` - will hold the app password itself for LDAP
+* `random` - auxiliary virtual attribute, a flag denoting whether to generate
+  the password
+* `randompassword` - auxiliary virtual attribute, stores the generated password
+  for display in the POST callback of the modification command
+* `managedby` - backlink to the owning user (that created it); it will be
+  automatically filled and maintained with respect to referential integrity;
+  search attribute; inspiration - otptoken ownership
+
+The definition of the new object and its attributes holding app passwords and
+related information, as well as operations with them - add, list, delete - will
+be contained in a new file - `ipaserver/plugins/apppw.py`.
+
+It will be possible to add app passwords only by generating them - FreeIPA's
+existing mechanism for generating passwords will be used for this.
+
+There will be no option to reset an app password.
```

Comment:
I think you will end up needing two IPA plugins to manage things. 

One plugin to manage the applications (add/delete/mod/show/find) and one to manage the passwords for those applications.

Reply:
Additionally, what does a user do if they've forgotten their app-specific password if there is no administrative reset? Do they log in using their main password and delete and re-add it?

Reply:
Regarding the [first post](https://github.com/freeipa/freeipa/pull/4061#discussion_r440276475): the user will be able to add an app password for whatever application; he/she will even set the app name manually. As you wrote [below](https://github.com/freeipa/freeipa/pull/4061#issuecomment-644243348), this design should be sufficient. Managing apps and passwords for them separately would also increase the complexity of the implementation, as well as of the management itself.

Regarding the [second post](https://github.com/freeipa/freeipa/pull/4061#discussion_r440277116): yes. This is easier to implement, and it was concluded [above](https://github.com/freeipa/freeipa/pull/4061#discussion_r396445912) that it is a way to go.

Reply:
Sure, it makes sense to me now. Typically creation is more admin-oriented but it completely makes sense to give the users some free reign in this situation.